### PR TITLE
feat: `onClickOutside`

### PR DESCRIPTION
**Pull request by sviripa**
This PR implements `useClickOutside` requested in #37

### CODE REVIEW

**Comment by TGlide** [comment]
@sviripa our internals have changed quite a bit since this PR. Sorry for the long time to review!  

Would you like to refactor it (maybe to a class too), or should I close it?

**Comment by sviripa** [comment]
@TGlide I'll refactor this one, no worries!

**Comment by huntabyte** [comment]
Hey @sviripa, are you still planning to refactor this, or should one of us push it over the finish line?

**Comment by sviripa** [comment]
@huntabyte @TGlide I just pushed the changes to make `useClickOutside` to use the new internals. Sorry for the delay

File: packages/runed/src/lib/utilities/useClickOutside/useClickOutside.svelte.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+};
+
+/**
+ * Accepts a box which holds a container element and callback function.
+ * Invokes the callback function when the user clicks outside of the
+ * container.
+ *
+ * @returns an object with start and stop functions
+ *
+ * @see {@link https://runed.dev/docs/functions/use-click-outside}
```
Comment (TGlide) [inline packages/runed/src/lib/utilities/useClickOutside/useClickOutside.svelte.ts:17]:
```suggestion
 * @see {@link https://runed.dev/docs/utilities/use-click-outside}
```

File: packages/runed/src/lib/utilities/useClickOutside/useClickOutside.svelte.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ */
+export function useClickOutside<T extends Element>(
+	container: MaybeGetter<T | undefined>,
+	callback: () => void
+): ClickOutside {
+	let isEnabled = $state<boolean>(true);
```
Comment (TGlide) [inline packages/runed/src/lib/utilities/useClickOutside/useClickOutside.svelte.ts:23]:
```suggestion
	let isEnabled = $state(true);
```

File: packages/runed/src/lib/utilities/useClickOutside/useClickOutside.svelte.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	const el = $derived<T | undefined>(extract(container));
```
Comment (TGlide) [inline packages/runed/src/lib/utilities/useClickOutside/useClickOutside.svelte.ts:24]:
```suggestion
	const el = $derived(extract(container));
```

File: packages/runed/src/lib/utilities/useClickOutside/useClickOutside.svelte.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+	function start() {
+		isEnabled = true;
+	}
+
+	function stop() {
+		isEnabled = false;
+	}
```
Comment (TGlide) [inline packages/runed/src/lib/utilities/useClickOutside/useClickOutside.svelte.ts:32]: I don't think these are needed. They can handle it on the event listener

File: packages/runed/src/lib/utilities/useClickOutside/useClickOutside.svelte.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+	function handleClick(event: MouseEvent) {
+		if (event.target && !el?.contains(event.target as Node)) {
+			callback();
+		}
+	}
+
+	watch([() => el, () => isEnabled], ([currentEl, currentIsEnabled]) => {
+		if (currentEl && currentIsEnabled) {
+			window.addEventListener("click", handleClick);
```
Comment (TGlide) [inline packages/runed/src/lib/utilities/useClickOutside/useClickOutside.svelte.ts:42]: You can use the `useEventListener` utility!

File: packages/runed/src/lib/utilities/useClickOutside/useClickOutside.svelte.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+	function stop() {
+		isEnabled = false;
+	}
+
+	function handleClick(event: MouseEvent) {
+		if (event.target && !el?.contains(event.target as Node)) {
+			callback();
+		}
+	}
```
Comment (TGlide) [inline packages/runed/src/lib/utilities/useClickOutside/useClickOutside.svelte.ts:38]:
```suggestion
	function handleClick(event: MouseEvent) {
		if (!event.target || el?.contains(event.target as Node)) return
		callback();
	}
```

**Review by TGlide — DISMISSED** [review]
Some small changes needed :)

**Comment by Hugos68** [comment]
I also think we should use `document` for event listeners like `click` instead of `window` as `document` catches the `click` event just as good and is closer to the source in the DOM tree so it makes more sense.

**Comment by sviripa** [comment]
@Hugos68 could you please provide an example of how `node.contains` approach does not work for dialog tags? I tested the utility with the dialog tag locally and it behaved as expected

**Comment by Hugos68** [comment]
@sviripa Sure! [REPL](https://svelte-5-preview.vercel.app/#H4sIAAAAAAAACqVUwY6bMBD9lYm1Ekai5J4AVVX11mo_oPRA7CHrXWMje0hbsfx7ZRyyNNntHnpA4JnnNzNvnhhZqzR6tvs-MtN0yHbsU9-zjNHvPhz8CTUhy5i3gxMhUnjhVE9VbWpSXW8dwQiDx89aiaf7gbySCBO0znaQ5FsRwh9sjOeRLn_0yT7cD49GAqkabY9Qwp2nhpCbQes0IExNwhpPYHs0UAJPoaxgnO-SaoFv4tV0idXkkAZnIj1N8RVBuX-wP79Z2Wg-k8_ZSwWhrcf_LjGz_E1_pQ2PBSI-i2WXWe-wbVEQD9OGWLF9EdsUh4HIGrBm1rQcA2qq7ns0xTbmIu6s5kEZuaMH5csxRqYqal5s47liGeusVK1CyXbkBpyyiwte29uLLR792hJRwGsPlMCpcUekDESj9aERT2tpw95Rh51LdOqE8ozm6aLGZfdx4ECIa4bLglDD8zNs8ISG8sgSAqg_5sIaapTxfJ1M0zVFTQChkNWYa3vkiTJz_1GkTZLu19ir7a8csDR84TmL9yrRoghPVzaaPxYP8DdGver9X91IK4YujN1I-SXM_1V5QoOOJ7OiSbZou24tMsJN_RWfw86e8B3Km5amUGWeE3-99d_Y33ryx_QHQfaQHaUEAAA=)

In the REPL I open the dialog by default for demo purposes and when you try to close it you'll see "inside dialog!" get logged.
This is because the `::backdrop` isn't a seperate element but actually a pseudo element of the `dialog` element. To solve this we need to check the `rect` of the `dialog` element so we can distinguish if a click was actually on the element or one of it's pseudo elements.

Here is a visualisation I just made:
image

**Comment by Hugos68** [comment]
Also, you may edit your original comment to `Closes #37` as this will trigger github to auto close the issue when this is merged.

**Comment by sviripa** [comment]
@Hugos68 Thanks for your comments and examples 🤝 

@TGlide Please take a look when you have time

**Comment by huntabyte** [comment]
@TGlide want to revisit this to see if it's g2g based on your previous feedback?

**Comment by huntabyte** [comment]
And perhaps we rename this is `onClickOutside`?

**Comment by huntabyte** [comment]
TODO: investigate what happens when a click occurs inside of an `iframe` and handle that accordingly.

**Comment by huntabyte** [comment]
Closing in favor of #190 for deployment

**Pull request closed without merging**